crates/accelerate/src/unitary_synthesis.rs

```diff
@@ -0,0 +1,1094 @@
+                        Ok(UnitarySynthesisReturnType::DAGType(Box::new(synth_dag)))
+                    }
+                }
+            }
+        }
+    }
+}
+
+// special-case synth function for the TwoQubitBasisDecomposer
+// used in `run_2q_unitary_synthesis`
+fn synth_su4_no_dag(
+    py: Python<'_>,
+    su4_mat: &Array2<Complex64>,
+    decomposer_2q: &DecomposerElement,
+    preferred_direction: Option<bool>,
+    approximation_degree: Option<f64>,
+) -> PyResult<UnitarySynthesisReturnType> {
+    let is_approximate = approximation_degree.is_none() || approximation_degree.unwrap() != 1.0;
+    let synth = if let DecomposerType::TwoQubitBasisDecomposer(decomp) = &decomposer_2q.decomposer {
+        decomp.call_inner(su4_mat.view(), None, is_approximate, None)?
+    } else {
+        panic!("synth su4 no dag should only be called for TwoQubitBasisDecomposer")
+    };
+
+    let sequence = TwoQubitUnitarySequence {
+        gate_sequence: synth.clone(),
+        decomp_gate: Some(decomposer_2q.decomp_gate.clone()),
+        decomp_gate_params: decomposer_2q.decomp_gate_params.clone(),
+    };
+
+    //synth_direction is calculated in terms of logical qubits
+    match preferred_direction {
+        None => Ok(UnitarySynthesisReturnType::TwoQSequenceType(sequence)),
+        Some(preferred_dir) => {
+            let mut synth_direction: Option<SmallVec<[u8; 2]>> = None;
+            for (gate, _, qubits) in synth.gates {
+                if gate.is_none() || gate.unwrap().name() == "cx" {
+                    synth_direction = Some(qubits);
+                }
+            }
+
+            match synth_direction {
+                None => Ok(UnitarySynthesisReturnType::TwoQSequenceType(sequence)),
+                Some(synth_direction) => {
+                    let synth_dir = match synth_direction.as_slice() {
+                        [0, 1] => true,
+                        [1, 0] => false,
+                        _ => panic!(),
+                    };
+                    if synth_dir != preferred_dir {
+                        reversed_synth_su4(py, su4_mat, decomposer_2q, approximation_degree)
+                    } else {
+                        Ok(UnitarySynthesisReturnType::TwoQSequenceType(sequence))
+                    }
+                }
+            }
+        }
+    }
+}
+
+// generic synth function for 2q gates (4x4) called from synth_su4
+fn reversed_synth_su4(
+    py: Python<'_>,
+    su4_mat: &Array2<Complex64>,
+    decomposer_2q: &DecomposerElement,
+    approximation_degree: Option<f64>,
+) -> PyResult<UnitarySynthesisReturnType> {
+    let is_approximate = approximation_degree.is_none() || approximation_degree.unwrap() != 1.0;
+    let mut su4_mat_mm = su4_mat.clone();
+
+    // Swap rows 1 and 2
+    let row_1 = su4_mat_mm.slice(s![1, ..]).to_owned();
+    let row_2 = su4_mat_mm.slice(s![2, ..]).to_owned();
+    su4_mat_mm.slice_mut(s![1, ..]).assign(&row_2);
+    su4_mat_mm.slice_mut(s![2, ..]).assign(&row_1);
```

Comment:
I think you can do something like:

```suggestion
    let (mut row_1, mut row_2) = su4_mat_mm.multi_slice_mut((s![1, ..], s![2, ..]));
    azip!((x in &mut row_1, y in &mut row2) (*x, *y) = (*y, *x))
```